consensus/istanbul/core/handler_test.go

```diff
@@ -464,6 +484,248 @@ func TestCore_handlerMsg(t *testing.T) {
+		validators      = mockBackend.Validators(lastBlock)
+		proposer        = validators.GetProposer()
+		proposerKey     = validatorKeyMap[proposer.Address()]
+	)
+
+	// Start istanbul core
+	istConfig := istanbul.DefaultConfig
+	istConfig.ProposerPolicy = istanbul.WeightedRandom
+	coreProposer := New(mockBackend, istConfig).(*core)
+	coreA := New(mockBackend, istConfig).(*core)
+	coreB := New(mockBackend, istConfig).(*core)
+	require.Nil(t,
+		coreProposer.Start(),
+		coreA.Start(),
+		coreB.Start())
+	defer coreProposer.Stop()
+	defer coreA.Stop()
+	defer coreB.Stop()
+
+	// make two groups
+	// the number of group size is (numValidators-1/2) + 1
+	// groupA consists of proposer, coreA, unnamed node(s)
+	// groupB consists of proposer, coreB, unnamed node(s)
+	subList := validators.SubList(lastBlock.Hash(), coreProposer.currentView())
+	groupA, groupB := splitSubList(subList, (numValidators-1)/2, proposer.Address())
+	groupA = append(groupA, proposer)
+	groupB = append(groupB, proposer)
+
+	// Step 1 - the malicious proposer generates two blocks
+	proposalA, err := genBlockParams(lastBlock, proposerKey, 0, 0, 1)
+	assert.Nil(t, err)
+
+	proposalB, err := genBlockParams(lastBlock, proposerKey, 1000, 10, 1)
+	assert.Nil(t, err)
+
+	// Shortcut for sending message `proposal` to core `c`
+	sendMessages := func(state uint64, proposal *types.Block, CNList []istanbul.Validator, c *core) {
+		for _, val := range CNList {
+			valKey := validatorKeyMap[val.Address()]
+			if state == msgPreprepare {
+				istanbulMsg, _ := genIstanbulMsg(state, lastBlock.Hash(), proposal, proposer.Address(), valKey)
+				err = c.handleMsg(istanbulMsg.Payload)
+			} else {
+				istanbulMsg, _ := genIstanbulMsg(state, lastBlock.Hash(), proposal, val.Address(), valKey)
+				err = c.handleMsg(istanbulMsg.Payload)
+			}
+			if err != nil {
+				t.Logf("handleMsg error: %s", err)
+			}
+		}
+	}
+	// Step 2 - exchange consensus messages inside each group
+
+	// the proposer sends two different blocks to each group
+	// each group receives a block and handles the message
+	// when chain split occurs, their states become StateCommitted
+	// otherwise, their states stay StatePreprepared
+	sendMessages(msgPreprepare, proposalA, groupA, coreA)
+	sendMessages(msgPrepare, proposalA, groupA, coreA)
+	if coreA.state.Cmp(StatePrepared) == 0 {
+		sendMessages(msgCommit, proposalA, groupA, coreA)
+	}
+
+	sendMessages(msgPreprepare, proposalB, groupB, coreB)
+	sendMessages(msgPrepare, proposalB, groupB, coreB)
+	if coreB.state.Cmp(StatePrepared) == 0 {
+		sendMessages(msgCommit, proposalB, groupB, coreB)
+	}
+
+	return coreA.state, coreB.state
+}
+
+// TestCore_chainSplit tests whether a chain split occurs in a certain conditions:
+// 1) the number of validators does not consist of 3f+1;
+//     e.g. if the number of validator is 5, it consists of 3f+2 (f=1)
+// 2) the proposer is malicious; it sends two different blocks to each group
+func TestCore_chainSplit(t *testing.T) {
```

Comment:
it looks good to me for now, but i think `chainsplit` will be a false test case  when https://github.com/klaytn/klaytn/issues/1403 is applied. Could you change the tc?

My Suggestion is increasing the number of `f` beyond the pbft assumption to fork.

Reply:
I think `func simulateChainSplit(t *testing.T, numTotalValidators, numMalValidators int)` form would be better since we can generalize malicious nodes tests. How do you think?